[ty] Add support for narrowing on tuple match cases (#25493)
astral-sh/ruff: crates/ty_python_semantic/resources/mdtest/narrow/match.md

Reviewer: One missing test case here is something like this:

```py
def f(x: tuple[int | str]) -> int:
    match x:
        case [int()]:
            return 42
        case [str()]:
            return 42
        case _:
            reveal_type(x)
```

We should narrow the type to `tuple[Never]` in that final branch, which simplifies immediately to `Never`. If we did so, we would not emit the spurious `invalid-return-type` annotation that we currently do.

Only mypy and pyright currently handle this correctly -- pyrefly, zuban and pycroscope all have the same false positive as we currently do on your branch (multiplay gist: 70a3fe08657bb858f231d66358b7f3e6).

I don't think we should try to fix this in this PR, but it would be great to add the test with a TODO comment

Reviewer: It doesn't look like we yet have any tests for the `& ~str & ~bytes & ~bytearray` part of the intersection. E.g.

```py
def f(x: str | tuple[int, int]):
    match x:
        case (a, b):
            reveal_type(a)  # revealed: @Todo
            reveal_type(b)  # revealed: @Todo
        case _:
            reveal_type(x)  # revealed: str
```

We wouldn't get the correct answer in the second branch there if it weren't for the `& ~str & ~bytes & ~bytearray` part of the intersection. We should add similar tests for unions with `bytes` and `bytearray` too

Author: I think this part doesn't work yet since we don't support narrowing on the assigned names?

Author: (Gah, sorry, I missed the `@Todo` in your own example...)

Reviewer: it's the `revealed: str` in the second branch that already works on your PR! and it wouldn't work unless we intersected the type of `x` with `& ~str & ~bytes & ~bytearray` in the first branch

        case int(), str():
            # revealed: Sequence[object] & <Protocol with members '__getitem__', '__len__'> & ~str & ~bytes & ~bytearray
            reveal_type(value)
            reveal_type(len(value))  # revealed: Literal[2]
            reveal_type(value[0])  # revealed: int
            reveal_type(value[1])  # revealed: str

def test_match_empty_object_sequence(value: object) -> None:
    match value:
        case []:
            # revealed: Sequence[object] & <Protocol with members '__len__'> & ~str & ~bytes & ~bytearray
            reveal_type(value)
            reveal_type(len(value))  # revealed: Literal[0]

def test_match_singleton_object_sequence(value: object) -> None:
    match value:
        case [int()]:
            # revealed: Sequence[object] & <Protocol with members '__getitem__', '__len__'> & ~bytearray & ~bytes
            reveal_type(value)
            reveal_type(len(value))  # revealed: Literal[1]
            reveal_type(value[0])  # revealed: int

def test_match_prefix_star_object_sequence(value: object) -> None:
    match value:
        case [int(), *rest]:
            reveal_type(value)  # revealed: Sequence[object] & ~str & ~bytes & ~bytearray

>             reveal_type(len(value))  # revealed: int

Reviewer: we obviously can't intersect with a protocol that has a precise length here, because of the wildcard. But we could intersect with a protocol that has a refined `__getitem__` method, I think?

```py
class P(Protocol):
    @overload
    def __getitem__(self, key: Literal[0], /) -> int: ...
    @overload
    def __getitem__(self, key: int, /) -> object: ...
```

then we would infer `int` for `reveal_type(x[0])`

# Exact patterns are commonly used to validate parsed data before using it.

def unwrap_single_int(value: object) -> int | None:
    match value:
        case [int()]:
            return value[0]
    return None

# This deliberately accepts a small unsoundness: sequence matching can inspect
# values yielded by `__iter__`, while a later indexed read calls `__getitem__`.
# Custom `Sequence` implementations and tuple subclasses can override those
# methods inconsistently. We assume conventional container behavior so exact
# patterns remain useful for validating parsed data before indexing it.
def accepted_unsound_indexed_reconstruction(value: Sequence[object]) -> int | None:
    match value:
        case [int()]:
            return value[0]

>     return None

Reviewer: these two tests look the same, and I'm not sure I understand the comment in between them

> def normalize_counted_label(value: object | None) -> str | None:

Reviewer: `object | None` just simplifies immediately to `object` (`None` is an instance of `object`, like everything else)

```suggestion
def normalize_counted_label(value: object) -> str | None:
```

    match value:
        case [int(), str()]:
            return value[1].upper() * value[0]
    return None

def test_match_exact_tuple_sequence(subj: tuple[int | str, int | str]) -> None:
    match subj:
        case x, str():
            # revealed: tuple[int | str, int | str] & <Protocol with members '__getitem__', '__len__'>
            reveal_type(subj)
        case y:
            # revealed: tuple[int | str, int | str] & ~<Protocol with members '__getitem__', '__len__'>

>             reveal_type(subj)

Reviewer: do we want to add TODOs here for how ideally we might simplify the first intersection to `tuple[int | str, str]` and the second to `tuple[int | str, int]`? And do we want to add some more `reveal_type` calls to show that we already infer the correct types from indexing now, even though we don't simplify the intersections yet?

def test_match_singleton_object_sequence(value: object) -> None:
    match value:
        case [int()]:
            # revealed: Sequence[object] & <Protocol with members '__getitem__', '__len__'> & ~bytearray & ~bytes
            reveal_type(value)
            reveal_type(len(value))  # revealed: Literal[1]
            reveal_type(value[0])  # revealed: int

def test_match_prefix_star_object_sequence(value: object) -> None:
    match value:
        case [int(), *rest]:
            # revealed: Sequence[object] & <Protocol with members '__getitem__'> & ~str & ~bytes & ~bytearray
            reveal_type(value)
            reveal_type(len(value))  # revealed: int
            reveal_type(value[0])  # revealed: int
            reveal_type(value[1])  # revealed: object

def test_match_prefix_and_suffix_star_object_sequence(value: object) -> None:
    match value:
        case [int(), *rest, str()]:
            # revealed: Sequence[object] & <Protocol with members '__getitem__'> & ~str & ~bytes & ~bytearray
            reveal_type(value)
            reveal_type(value[0])  # revealed: int
            reveal_type(value[-1])  # revealed: str

>             reveal_type(value[1])  # revealed: object

Reviewer: this is so cool -- great job!

def test_match_prefix_star_known_sequence(value: Sequence[int | str]) -> None:
    match value:
        case [int(), *rest]:
            reveal_type(value[0])  # revealed: int
            reveal_type(value[1])  # revealed: int | str

def test_match_exact_tuple_sequence(subj: tuple[int | str, int | str]) -> None:
    match subj:
        case x, str():
            # TODO: This should simplify to `tuple[int | str, str]`.
            # revealed: tuple[int | str, int | str] & <Protocol with members '__getitem__', '__len__'>
            reveal_type(subj)
            reveal_type(subj[0])  # revealed: int | str

>             reveal_type(subj[1])  # revealed: str

Reviewer: ```suggestion
        case x, str():
            # TODO: This should simplify to `tuple[int | str, str]`.
            # revealed: tuple[int | str, int | str] & <Protocol with members '__getitem__', '__len__'>
            reveal_type(subj)
            reveal_type(subj[0])  # revealed: int | str
            reveal_type(subj[1])  # revealed: str
            first, second = subj
            reveal_type(first)  # revealed: int | str
            # TODO: this should reveal `str`
            reveal_type(second)  # revealed: int | str
```

        case y:
            # TODO: This should simplify to `tuple[int | str, int]`.
            # revealed: tuple[int | str, int | str] & ~<Protocol with members '__getitem__', '__len__'>
            reveal_type(subj)
            reveal_type(subj[0])  # revealed: int | str
            # TODO: This should reveal `int` once we simplify the negative
            # intersection above.
            reveal_type(subj[1])  # revealed: int | str

def test_match_exact_tuple_sequence_is_exhaustive(value: int | tuple[int, int]) -> int:
    match value:
        case int(value):
            return value
        case (left, right):
            return left + right
        case _:
            assert_never(value)

def test_match_exact_mutable_sequence_negative(value: list[int]) -> None:
    match value:
        case [int()]:
            pass
        case _:

>             # revealed: list[int] & ~<Protocol with members '__getitem__', '__len__'>
>             reveal_type(value)
> 
> def normalize_nested_record(value: object) -> tuple[None, int, int] | None:
>     match value:
>         case [None, [int()], {}]:
>             return value[0], value[1][0], len(value[2])
>     return None
> 
> def unwrap_number_or_label(value: object) -> int | str | None:
>     match value:
>         case [(int() | str()) as item]:
>             return value[0]
>     return None

Reviewer: the fact that we don't emit errors on the `return` statements shows us something, but not much, here -- if `value[0]`, `value[1][0]` and `value[2]` are all inferred as `Unknown`, the test would still pass. Maybe something like this?

```suggestion
def normalize_nested_record(value: object) -> tuple[None, int, int] | None:
    match value:
        case [None, [int()], {}]:
            ret = value[0], value[1][0], len(value[2])
            # revealed: tuple[None, int, int]
            reveal_type(ret)
            return ret
    return None

def unwrap_number_or_label(value: object) -> int | str | None:
    match value:
        case [(int() | str()) as item]:
            reveal_type(value[0])  # revealed: int | str
            return value[0]
    return None
```
